### PR TITLE
Block mount tester

### PR DESCRIPTION
**Pull request by carlbraganza**
```
/repovol/kastenhq/kubestr$ ./kubestr blockmount -h
Checks if volumes provisioned by a storage class can be mounted
in block mode.

The test works as follows:
- It dynamically provisions a volume of the given storage class.
- It then launches a pod with the volume mounted as a block device.
- If the pod is successfully created then the test passes. If it
  fails or times out then the test fails.

Usage:
  kubestr blockmount -s StorageClass [flags]

Flags:
  -c, --cleanup               Clean up the objects created by tool. (default true)
      --cleanup-only          Do not run the test, but just clean up resources left from a previous invocation.
  -h, --help                  help for blockmount
  -i, --image string          The container image used to create a pod.
  -n, --namespace string      The namespace used to run the check. (default "default")
  -u, --runAsUser int         Runs the block mount test pod with the specified user ID (int)
  -s, --storageclass string   The name of a Storageclass. (Required)
  -w, --wait-timeout uint32   Max time to wait for the test pod to become ready (default 60)

Global Flags:
  -e, --outfile string   The file where test results will be written
  -o, --output string    Options(json)
```

### CODE REVIEW

**Comment by carlbraganza** [comment]
Tested on AWS, GKE and Azure clusters with storage classes that both support and do not support block mode volumes.

**Review by bathina2 — COMMENTED** [review]
Couple comments

File: cmd/rootCmd.go [hunk 1]
```diff
@@ -97,6 +98,30 @@ var (
 			)
 		},
 	}
+
+	blockMountRunAsUser          int64
+	blockMountCleanup            bool
+	blockMountCleanupOnly        bool
+	blockMountWaitTimeoutSeconds uint32
+	blockMountCmd                = &cobra.Command{
+		Use:   "blockmount -s StorageClass",
```
[the remark this answers is not in the document]
Comment (carlbraganza) [reply cmd/rootCmd.go:107]: Changed, but it worked.

File: cmd/rootCmd.go [hunk 2]
```diff
@@ -306,3 +341,68 @@ func CsiPvcBrowse(ctx context.Context,
 	}
 	return err
 }
+
+func TestBlockMount(ctx context.Context, output, outfile,
```
[the remark this answers is not in the document]
Comment (carlbraganza) [reply cmd/rootCmd.go:345]: Fine. I'll use that for flags that can be passed in the argument type.

File: pkg/block/block_mount.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+func (a *BlockMountTesterArgs) Validate() error {
+	if a.KubeCli == nil || a.DynCli == nil || a.StorageClass == "" || a.Namespace == "" {
+		return fmt.Errorf("Require fields are missing. (KubeCli, DynCli, StorageClass, Namespace)")
+	}
+	return nil
+
+}
+
+// BlockMountTester tests if a storage class can provision volumes for block mounts.
+type BlockMountTester interface {
```
[the remark this answers is not in the document]
Comment (carlbraganza) [reply pkg/block/block_mount.go:40]: ok

**Comment by carlbraganza** [comment]
New usage:
```
/repovol/kastenhq/kubestr$ ./kubestr blockmount -h
Checks if volumes provisioned by a storage class can be mounted in block mode.

The checker works as follows:
- It dynamically provisions a volume of the given storage class.
- It then launches a pod with the volume mounted as a block device.
- If the pod is successfully created then the test passes.
- If the pod fails or times out then the test fails.

In case of failure, re-run the checker with the "-c" flag and examine the
failed PVC and Pod: it may be necessary to adjust the default values used for
the PVC size, the pod wait timeout, etc. Clean up the failed resources by
running the checker with the "--cleanup-only" flag.

Usage:
  kubestr blockmount [flags]

Flags:
      --cleanup-only          Do not run the checker, but just clean up resources left from a previous invocation.
  -h, --help                  help for blockmount
  -i, --image string          The container image used to create a pod.
  -n, --namespace string      The namespace used to run the check. (default "default")
  -c, --no-cleanup            Do not clean up the objects created by the check.
      --pvc-size string       The size of the provisioned PVC. (default "1Gi")
  -u, --runAsUser int         Runs the block mount check pod with the specified user ID (int)
  -s, --storageclass string   The name of a Storageclass. (Required)
  -w, --wait-timeout uint32   Max time in seconds to wait for the check pod to become ready (default 60)

Global Flags:
  -e, --outfile string   The file where test results will be written
  -o, --output string    Options(json)
```
Example of size flag:
```
/repovol/kastenhq/kubestr$ ./kubestr blockmount -s regional-disk --pvc-size=200Gi
Fetching StorageClass regional-disk ...
 -> Provisioner: kubernetes.io/gce-pd
Provisioning a Volume (200Gi) for block mode access ...
 -> Created PVC default/kubestr-blockmount-regional-disk-pvc (40ms)
Creating a Pod with a volumeDevice ...
 -> Created Pod default/kubestr-blockmount-regional-disk-pod
 -> Waiting at most 1m0s for the Pod to become ready ...
 -> The Pod is ready (23.399s)
Deleting Pod default/kubestr-blockmount-regional-disk-pod ...
 -> Deleted pod (33.088s)
Deleting PVC default/kubestr-blockmount-regional-disk-pvc ...
 -> Deleted PVC (63ms)
Block VolumeMode test:
  StorageClass (regional-disk) supports Block VolumeMode  -  OK
```

**Comment by carlbraganza** [comment]
UT in `pkg/block`:
```
=== RUN   TestBlockMountCheckerNew
=== RUN   TestBlockMountCheckerNew/args:empty
=== RUN   TestBlockMountCheckerNew/args:KubeCli
=== RUN   TestBlockMountCheckerNew/args:KubeCli-DynCli
=== RUN   TestBlockMountCheckerNew/args:KubeCli-DynCli-StorageClass
=== RUN   TestBlockMountCheckerNew/success
--- PASS: TestBlockMountCheckerNew (0.00s)
    --- PASS: TestBlockMountCheckerNew/args:empty (0.00s)
    --- PASS: TestBlockMountCheckerNew/args:KubeCli (0.00s)
    --- PASS: TestBlockMountCheckerNew/args:KubeCli-DynCli (0.00s)
    --- PASS: TestBlockMountCheckerNew/args:KubeCli-DynCli-StorageClass (0.00s)
    --- PASS: TestBlockMountCheckerNew/success (0.00s)
=== RUN   TestBlockMountCheckerPvcWaitForTermination
=== RUN   TestBlockMountCheckerPvcWaitForTermination/success
=== RUN   TestBlockMountCheckerPvcWaitForTermination/timeout
--- PASS: TestBlockMountCheckerPvcWaitForTermination (0.00s)
    --- PASS: TestBlockMountCheckerPvcWaitForTermination/success (0.00s)
    --- PASS: TestBlockMountCheckerPvcWaitForTermination/timeout (0.00s)
=== RUN   TestBlockMountCheckerCleanup
=== RUN   TestBlockMountCheckerCleanup/nothing-found
Deleting Pod namespace/kubestr-blockmount-sc-pod ...
 -> Deleted pod (0s)
Deleting PVC namespace/kubestr-blockmount-sc-pvc ...
 -> Deleted PVC (0s)
=== RUN   TestBlockMountCheckerCleanup/error-deleting-pod
Deleting Pod namespace/kubestr-blockmount-sc-pod ...
  Error deleting Pod namespace/kubestr-blockmount-sc-pod - (test error)
 -> Failed to delete Pod in 0s
Deleting PVC namespace/kubestr-blockmount-sc-pvc ...
 -> Deleted PVC (0s)
=== RUN   TestBlockMountCheckerCleanup/error-deleting-pvc
Deleting Pod namespace/kubestr-blockmount-sc-pod ...
 -> Deleted pod (0s)
Deleting PVC namespace/kubestr-blockmount-sc-pvc ...
  Error deleting PVC namespace/kubestr-blockmount-sc-pvc - (test error)
 -> PVC failed to delete in 0s
--- PASS: TestBlockMountCheckerCleanup (0.00s)
    --- PASS: TestBlockMountCheckerCleanup/nothing-found (0.00s)
    --- PASS: TestBlockMountCheckerCleanup/error-deleting-pod (0.00s)
    --- PASS: TestBlockMountCheckerCleanup/error-deleting-pvc (0.00s)
=== RUN   TestBlockMountCheckerMount
=== RUN   TestBlockMountCheckerMount/no-storage-class
Fetching StorageClass sc ...
 -> Failed to fetch StorageClass(sc): ( "sc" not found)
=== RUN   TestBlockMountCheckerMount/invalid-pvc-size
Fetching StorageClass sc ...
 -> Provisioner: provisioenr
 -> Invalid PVC size 10Q: (quantities must match the regular expression '^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$')
=== RUN   TestBlockMountCheckerMount/create-pvc-error
Fetching StorageClass sc ...
 -> Provisioner: provisioenr
Provisioning a Volume (1Gi) for block mode access ...
 -> Failed to provision a Volume (test error)
Deleting Pod namespace/kubestr-blockmount-sc-pod ...
 -> Deleted pod (0s)
Deleting PVC namespace/kubestr-blockmount-sc-pvc ...
 -> Deleted PVC (0s)
=== RUN   TestBlockMountCheckerMount/create-pod-error
Fetching StorageClass sc ...
 -> Provisioner: provisioenr
Provisioning a Volume (1Gi) for block mode access ...
 -> Created PVC namespace/kubestr-blockmount-sc-pvc (0s)
Creating a Pod with a volumeDevice ...
 -> Failed to create Pod (test error)
Deleting Pod namespace/kubestr-blockmount-sc-pod ...
 -> Deleted pod (0s)
Deleting PVC namespace/kubestr-blockmount-sc-pvc ...
 -> Deleted PVC (0s)
=== RUN   TestBlockMountCheckerMount/wait-for-pod-error
Fetching StorageClass sc ...
 -> Provisioner: provisioenr
Provisioning a Volume (1Gi) for block mode access ...
 -> Created PVC namespace/kubestr-blockmount-sc-pvc (0s)
Creating a Pod with a volumeDevice ...
 -> Created Pod namespace/kubestr-blockmount-sc-pod
 -> Waiting at most 0s for the Pod to become ready ...
 -> The Pod timed out (test error)
Deleting Pod namespace/kubestr-blockmount-sc-pod ...
 -> Deleted pod (0s)
Deleting PVC namespace/kubestr-blockmount-sc-pvc ...
 -> Deleted PVC (0s)
=== RUN   TestBlockMountCheckerMount/success-no-cleanup
Fetching StorageClass sc ...
 -> Provisioner: provisioenr
Provisioning a Volume (1Gi) for block mode access ...
 -> Created PVC namespace/kubestr-blockmount-sc-pvc (0s)
Creating a Pod with a volumeDevice ...
 -> Created Pod namespace/kubestr-blockmount-sc-pod
 -> Waiting at most 0s for the Pod to become ready ...
 -> The Pod is ready (0s)
--- PASS: TestBlockMountCheckerMount (0.00s)
    --- PASS: TestBlockMountCheckerMount/no-storage-class (0.00s)
    --- PASS: TestBlockMountCheckerMount/invalid-pvc-size (0.00s)
    --- PASS: TestBlockMountCheckerMount/create-pvc-error (0.00s)
    --- PASS: TestBlockMountCheckerMount/create-pod-error (0.00s)
    --- PASS: TestBlockMountCheckerMount/wait-for-pod-error (0.00s)
    --- PASS: TestBlockMountCheckerMount/success-no-cleanup (0.00s)
PASS
	github.com/kastenhq/kubestr/pkg/block	coverage: 100.0% of statements
ok  	github.com/kastenhq/kubestr/pkg/block	0.166s	coverage: 100.0% of statements
```

**Review by bathina2 — APPROVED** [review]

**Pull request merged**